### PR TITLE
feat: apache Pulsar bindings

### LINKED ISSUE #164: Pulsar bindings for AsyncAPI
#### Reason/Context
Apache® Pulsar™ is a distributed, open source pub-sub messaging and streaming platform.
Ability to use AsyncApi spec to manage (versioning, schema, bindings, code generation, etc) the platform will be invaluable to organisations.

#### Description
A new AsyncAPI binding for Pulsar to be provided, possibly mirroring the same approach used for all other bindings

### PR DESCRIPTION
**Pull request by VisualBean**
Adds Server and channel bindings for Apache Pulsar

**Related issue(s)**
Resolves #164

### CODE REVIEW

**Comment by VisualBean** [comment]
Is it allowed to collaborate on a PR here, or should we do it on the Fork? Id really like some early comments, before we fully go for it, but if it is against the contributing guidelines, please close :)

**Comment by austek** [comment]
sure we can collaborate, I'm on AsyncApi Slack

**Comment by fmvilas** [comment]
It's completely fine to collaborate here if you can give @austek permissions to edit this PR.

File: pulsar/README.md
Comment (VisualBean) [inline pulsar/README.md:48]: Perhaps adding `namespace` to the channel binding object

File: pulsar/README.md
Comment (VisualBean) [inline pulsar/README.md:19]: Perhaps adding the Pulsar Tenant here would be a nice addition, as it cannot really be inferred from the Server.Url

**Comment by VisualBean** [comment]
One big question is,
Pulsar has tenants, as the "top" of the hierarchy.

In order to Produce to a given topic, a uri that looks like this is needed.
`{persistent|non-persistent}://tenant/namespace/topic`

In the current iteration, ive attached Tenant to the Server binding, to try and reduce repitition, but what kind of fleksibility are we looking for in the bindings?

In the usecases im looking at, AsyncApi documents are per tenant. 
@austek for additional input.

**Review by heesung-sn — CHANGES_REQUESTED** [review]

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json:18]: I saw another binding that used strings for this,which is why i went with strong, but yes, as pulsar only has 2 it should be fine.

The argument for string could be, that it is easier to infer the topic string when connecting. Instead of having to know the vendor specifics that persistence means "persistent", but rather a topic looks like 
`{persistence}/{tenant}/{namspace}/{channel}` if you know what I mean. Which i personally think is a bit closer to what we want out of the binding.

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json:14]: I think server is too high level for namespaces. Maybe even for tenant.

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json:26]: Again, boolean is not true to the vendor specifics in this case. 

When producing to a topic (that i don't myself own), i would be very interested in what the parameters for compaction are. 

Especially if we look at usage with auto generating clients/servers.

File: pulsar/json_schemas/server.json
Comment (VisualBean) [inline pulsar/json_schemas/server.json:32]:
```suggestion
      "tenant": "contoso",
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:44]: Should it be required?

**Review by fmvilas — COMMENTED** [review]
Left a comment. I have no idea about Pulsar so we have to fully trust you on this folks 😄 

Also, would you be up for becoming the code owner of this binding? You would be listed at https://github.com/asyncapi/bindings/blob/master/CODEOWNERS and would become a [TSC member](https://asyncapi.com/community/tsc).

File: pulsar/README.md
Comment (VisualBean) [reply pulsar/README.md:44]: Yes. I was unsure about how to set that up specifically and what the recommendation was, but essentially yes. This particular one cannot be defaulted.

**Comment by VisualBean** [comment]
> Left a comment. I have no idea about Pulsar so we have to fully trust you on this folks 😄
> 
> Also, would you be up for becoming the code owner of this binding? You would be listed at https://github.com/asyncapi/bindings/blob/master/CODEOWNERS and would become a [TSC member](https://asyncapi.com/community/tsc).

I would love to.

File: pulsar/README.md
Comment (fmvilas) [reply pulsar/README.md:44]: Awesome. So then I'd just mark it as required:

```suggestion
`persistence` | String | **Required**. Persistence of the topic in Pulsar. Its value MUST be either `persistent` or `non-persistent`. |
```

Comment (VisualBean) [reply pulsar/README.md:44]: Awesome. I'll go over them and explicitly write required/optional then. Thanks!

File: pulsar/json_schemas/channel.json
Comment (heesung-sn) [inline pulsar/json_schemas/channel.json:50]: Do we need to define the other pulsar components here?
`partition`
`ttl`,
`dead-letter-queue`?

**Review by heesung-sn — COMMENTED** [review]
Are we planning to add Pulsar bindings for `message` and `operation` AsyncAPI components?

I think the followings are also important to define in the pulsar bindings:
`AccessMode`
https://pulsar.apache.org/docs/2.10.x/concepts-messaging#access-mode

`Subscription Type`
https://pulsar.apache.org/docs/2.10.x/concepts-messaging#subscription-types

`Subscription Mode
https://pulsar.apache.org/docs/2.10.x/concepts-messaging#subscription-modes`

Although we can add more Pulsar features in the next iterations of binding spec, it would be great if we can capture the major features in the first version. FYI, I notified the Pulsar community about this work in this email thread, https://lists.apache.org/thread/q8hlg27c1yh94w73j53dy2y54yyjdt62

**Comment by VisualBean** [comment]
> Are we planning to add Pulsar bindings for `message` and `operation` AsyncAPI components?
> 
> I think the followings are also important to define in the pulsar bindings:
> `AccessMode`
> https://pulsar.apache.org/docs/2.10.x/concepts-messaging#access-mode
> 
> `Subscription Type`
> https://pulsar.apache.org/docs/2.10.x/concepts-messaging#subscription-types
> 
> `Subscription Mode`
> https://pulsar.apache.org/docs/2.10.x/concepts-messaging#subscription-modes
> 
> Although we can add more Pulsar features in the next iterations of binding spec, it would be great if we can capture the major features in the first version. FYI, I notified the Pulsar community about this work in this email thread, https://lists.apache.org/thread/q8hlg27c1yh94w73j53dy2y54yyjdt62
> 
> 

You are very welcome to put suggestions, but i definitely think the subscription specific things, fit the operations binding. 
I couldn't immediately think of anything that fits a message binding, but if there are, let's get them going as well.

**Comment by heesung-sn** [comment]
I think we could first try to agree on the definitions of server, channel, operation, and message bindings in the Pulsar context.

From my understanding, we are going with the below proposal in this PR?

server: tenant level broker-admin configurations. 
channel: namespace/topic level broker-admin configurations.
operation: pub/sub client level configurations.
message: message level configs(?) -- not sure about this message bindings for pulsar(perhaps TBD)

**Comment by VisualBean** [comment]
> server: tenant level broker-admin configurations.
> channel: namespace/topic level broker-admin configurations.
> operation: pub/sub client level configurations.
> message: message level configs(?) -- not sure about this message bindings for pulsar(perhaps TBD)

Yes, that is my take as well, although, tenant could feasibly be channel level, im a bit up in the air about that one.

For the first version, I wanted to leave out all of the client dynamic configuration things, so essentially trying to stick to what can be set through the REST API, if that makes sense. 

Having thought a bit about it, I think dead-letter-queue could make sense, as well as some other producer specifics, but perhaps those can be in a future version, as they are dynamically configurable by the clients producing to the topics in question, and as such, are much harder to determine "accurate" configuration of.

**Comment by heesung-sn** [comment]
> > server: tenant level broker-admin configurations.
> > channel: namespace/topic level broker-admin configurations.
> > operation: pub/sub client level configurations.
> > message: message level configs(?) -- not sure about this message bindings for pulsar(perhaps TBD)
> 
> Yes, that is my take as well, although, tenant could feasibly be channel level, im a bit up in the air about that one.
> 
> For the first version, I wanted to leave out all of the client dynamic configuration things, so essentially trying to stick to what can be set through the REST API, if that makes sense, as that configuration is bound by the cluster, and not by the clients.
> 
> Having thought a bit about it, I think dead-letter-queue could make sense, as well as some other producer/consumer specifics, but perhaps those can be in a future version, as they are dynamically configurable by the clients producing to the topics in question, and as such, are much harder to determine "accurate" configuration of.

Ok. this direction makes sense to me too. We could stick to the broker-admin configs first in this PR.

Also, server:  `cluster`/`tenant` level broker-admin configs make sense to me too in this first version.

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json:54]: I have added TTL, good eye. Completely neglected it.

File: pulsar/json_schemas/server.json
Comment (VisualBean) [reply pulsar/json_schemas/server.json:13]: I will add replication 👍 to the channel binding,

Comment (VisualBean) [reply pulsar/json_schemas/server.json:13]: Take a look, ive added geo-replication to the channel binding.

File: pulsar/README.md
Comment (heesung-sn) [inline pulsar/README.md:80]: Don't we need to remove this retention object?

Does it appear that we want to define the server, channel, operation, and messaging binding fields here in the readme?

File: pulsar/json_schemas/channel.json
Comment (heesung-sn) [inline pulsar/json_schemas/channel.json:5]: Let's define that this `channel` defines Pulsar namespace/topic level admin configs here in the description to clarify.

File: pulsar/json_schemas/channel.json
Comment (heesung-sn) [inline pulsar/json_schemas/channel.json:70]: missing `ttl` here in the example.

Comment (heesung-sn) [reply pulsar/json_schemas/channel.json:70]: missing `namespace` here in the example.

File: pulsar/json_schemas/channel.json
Comment (heesung-sn) [inline pulsar/json_schemas/channel.json:13]: Don't we need `topic-name` here?

File: pulsar/README.md
Comment (VisualBean) [reply pulsar/README.md:80]: If you look around at the other bindings, a similar structure has been used. As well as a definition for certain objects (like our retention object)

File: pulsar/json_schemas/server.json
Comment (heesung-sn) [inline pulsar/json_schemas/server.json:24]: can't we add `cluster` here?

File: pulsar/json_schemas/server.json
Comment (heesung-sn) [inline pulsar/json_schemas/server.json:5]: Let's clarify that this `server` binding specify Pulsar `cluster`/`tenant` level config here too.

Comment (VisualBean) [reply pulsar/json_schemas/server.json]: Done

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json:71]: Fixed

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json]: Done

**Comment by lin-zhao** [comment]
Do we also need binding for messages? If my understanding is correct, message binding is what's needed for defining schemas right?

**Comment by austek** [comment]
How about schema, shouldn't it be part of the API spec?

**Comment by VisualBean** [comment]
> Do we also need binding for messages? If my understanding is correct, message binding is what's needed for defining schemas right?

No, you can specify the schema as part of the message, in the operation, in the asyncapi document. 

I don't see a need for adding it to the binding.

File: pulsar/README.md
Comment (VisualBean) [inline pulsar/README.md:63]:
```suggestion
        compaction: 1000
        geo-replication:
          - us-east1
          - us-west1
```

**Comment by VisualBean** [comment]
@fmvilas I think we are somewhat stable at this point. 
How do we procede?

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:18]: Let's change the alignment of the description field to "left-aligned".

```suggestion
---|:---:|:---|
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:42]: Same as above:

```suggestion
---|:---:|:---|
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:44]: Can we move this to the top of the table? Since it's the only required property it may help make it more visible.

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:43]:
```suggestion
`namespace` | String | **Optional**. The namespace the channel is associated with. If omitted, "public" MUST be assumed. |
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:44]:
```suggestion
`persistence` | String | **Required**. Persistence of the topic in Pulsar. It MUST be either `persistent` or `non-persistent`. |
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:48]:
```suggestion
`ttl` | Integer | **Optional**. Message time-to-live in seconds. |
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:49]:
```suggestion
`deduplication` | Boolean | **Optional**. Message deduplication. When true, it ensures that each message produced on Pulsar topics is persisted to disk only once. |
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:84]: If this object is only used in the Channel Binding Object, I'd move it under that section. And would use an "H3", i.e.:

```suggestion
### Retention Definition Object
```

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:90]: What does "by default" means here? These are required properties so they shouldn't have defaults, right? 🤔

File: pulsar/json_schemas/channel.json
Comment (fmvilas) [inline pulsar/json_schemas/channel.json:12]: If you want to allow extensions, then you should update the readme document where it says "This object MUST contain only the properties defined above." and instead change it to something like:

```markdown
This object MAY be extended with [Specification Extensions](https://www.asyncapi.com/docs/reference/specification/v2.5.0#specificationExtensions).
```

And change this block to:

```suggestion
    "patternProperties": {
      "^x-[\\w\\d\\.\\-\\_]+$": {
        "$ref": "https://raw.githubusercontent.com/asyncapi/spec-json-schemas/v4.0.0/schemas/2.5.0.json#/definitions/http://asyncapi.com/definitions/2.5.0/specificationExtension.json"
      }
    },
```

Although, I'm not sure this link is correct so I'm pinging @jonaslagoni just in case.

File: pulsar/json_schemas/channel.json
Comment (fmvilas) [inline pulsar/json_schemas/channel.json:33]: You can go even further and define it's an array of strings:

```suggestion
        "description": "A list of clusters the topic is replicated to.",
        "items": {
          "type": "string"
        }
```

**Review by fmvilas — CHANGES_REQUESTED** [review]
Looking great @VisualBean! Left a bunch of —mainly— editorial fixes. I'm pinging the other maintainers so they're aware too @derberg @dalelane @char0n @smoya.

File: pulsar/README.md
Comment (VisualBean) [reply pulsar/README.md:44]: You bet, makes more sense.

File: pulsar/README.md
Comment (VisualBean) [reply pulsar/README.md:90]: You are absolutely right, they should be optional, and the text changed.

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json:33]: Awesome, thanks a lot for all of the edits, ill move things around as suggested as well.

File: pulsar/json_schemas/channel.json
Comment (VisualBean) [reply pulsar/json_schemas/channel.json:12]: Looking at the other bindings, most of them has the text, along with the extensions.
AsyncAPI defines all bindings can be extended, so i figured the schema MUST include this.
and that the text then referred to "other" properties. 

But you are absolutely right.

File: pulsar/README.md
Comment (VisualBean) [reply pulsar/README.md]: Caught a copy paste mistake, with the defaulted namespace - this has been rectified, as this is also required.

File: pulsar/README.md
Comment (VisualBean) [reply pulsar/README.md]: Removed (By default)

**Review by derberg — COMMENTED** [review]
yo, left some editorial comments

➕ I have some other comments

- in `bindings` repo we always ask binding contributors if they want to become given binding maintainers. Of course, remember that you are not left alone, and main spec maintainers can always jump in and help if there is something that needs to be cleared out and consulted. So are you willing to maintain it? especially that I see there will be some followup PRs to add bindings to other objects.
- I assume you'd like to use January release window to get `pulsar` on the list of bindings in the main spec? We need to know as we need to start looking for volunteer to coordinate 2.6 release

anyway man, great PR 👏🏼 
also, looking at the number of comments, I'm guessing the Pulsar community was really looking forward to this one 💪🏼

File: pulsar/README.md
Comment (VisualBean) [reply pulsar/README.md]: You are 100% correct. Moved 👍.

**Comment by VisualBean** [comment]
> yo, left some editorial comments
> 
> ➕ I have some other comments
> 
>     * in `bindings` repo we always ask binding contributors if they want to become given binding maintainers. Of course, remember that you are not left alone, and main spec maintainers can always jump in and help if there is something that needs to be cleared out and consulted. So are you willing to maintain it? especially that I see there will be some followup PRs to add bindings to other objects.
> 
>     * I assume you'd like to use January release window to get `pulsar` on the list of bindings in the main spec? We need to know as we need to start looking for volunteer to coordinate 2.6 release
> 
> 
> anyway man, great PR 👏🏼 also, looking at the number of comments, I'm guessing the Pulsar community was really looking forward to this one 💪🏼

* I am very willing to maintain it.
* There might not be a need for operation and message bindings, we will see. Pulsar takes care of most of that behind the scenes, so I have yet to uncover an actual need for these, that arent already described by the asyncapi specification.
* January release would be perfect.

File: pulsar/README.md
Comment (fmvilas) [inline pulsar/README.md:17]: What about we have a table with the following headers instead:

* Field Name
* Type
* Required (yes/no)
* Description
* Default value (N/A when it's required)

Comment (VisualBean) [reply pulsar/README.md:17]: That works for me.
Glancing over the other bindings, there didn't appear to be a standard, but i kind of liked what the Kafka binding looked like

Comment (VisualBean) [reply pulsar/README.md]: Fixed. Have another look

Comment (fmvilas) [reply pulsar/README.md]: Yeah, bindings are still in diapers 😄 We'll slowly find a good pattern to follow.

**Review by fmvilas — DISMISSED** [review]
That's looking great mate 👏

**Comment by fmvilas** [comment]
Let's wait for other reviews \cc @derberg @dalelane @char0n

**Comment by derberg** [comment]
> I am very willing to maintain it.

Awesome! thanks so much. Then please update https://github.com/asyncapi/bindings/blob/master/CODEOWNERS in your PR 🙏🏼 

> January release would be perfect.

How much do you plan to engage with AsyncAPI community. Are you looking for something more than just pulsar binding?

I'm asking because the specification releases are always coordinated by a release coordinator. This is a volunteer role that anyone can pick up. I'm not gonna lie; it ain't easy and requires some work. You can [read more here](https://github.com/asyncapi/spec/blob/master/RELEASE_PROCESS.md). But the big advantage is that you learn a lot about how AsyncAPI is organized and tooling ecosystem and get in touch with different maintainers around. And yeah, we are always here to support you of course.

Give it a thought.

**Comment by VisualBean** [comment]
> > I am very willing to maintain it.
> 
> Awesome! thanks so much. Then please update https://github.com/asyncapi/bindings/blob/master/CODEOWNERS in your PR 🙏🏼
> 
> > January release would be perfect.
> 
> How much do you plan to engage with AsyncAPI community. Are you looking for something more than just pulsar binding?
> 
> I'm asking because the specification releases are always coordinated by a release coordinator. This is a volunteer role that anyone can pick up. I'm not gonna lie; it ain't easy and requires some work. You can [read more here](https://github.com/asyncapi/spec/blob/master/RELEASE_PROCESS.md). But the big advantage is that you learn a lot about how AsyncAPI is organized and tooling ecosystem and get in touch with different maintainers around. And yeah, we are always here to support you of course.
> 
> Give it a thought.

I am building a dotnet library that reads/writes asyncapi documents based on the objectmodel, which will go opensource very soon (hopefully) as well, so I am hoping to engage more, but am not ready to take on more work unfortunately.

**Review by dalelane — APPROVED** [review]
I'm not familiar with Pulsar, so this isn't a super-informed review - but as a bindings doc it looks fine to me 👍

File: pulsar/README.md
Comment (KhudaDad414) [inline pulsar/README.md:52]: It says that _if retention is specified, both fields are mandatory._ but they are optional in JSON Schema and in the table below, shouldn't they be `required` in Retention Definition Object?

Comment (VisualBean) [inline pulsar/README.md:52]:
```suggestion
The `Retention Definition Object` is used to describe the Pulsar [Retention](https://pulsar.apache.org/docs/cookbooks-retention-expiry/) policy.
```

**Review by KhudaDad414 — APPROVED** [review]
Schemas seems to match the descriptions 👍

**Review by fmvilas — APPROVED** [review]

**Comment by VisualBean** [comment]
Can someone rerun the sentiment analysis, looks like its failing due to a failure in the imported action
"Error: Unable to resolve action `someimportantcompany/github-actions-slack-message@v1`, unable to find version `v1`"

**Review by derberg — APPROVED** [review]

**Comment by fmvilas** [comment]
@dalelane can you please re-approve if you agree with the changes?

**Comment by fmvilas** [comment]
Congrats and thanks @VisualBean! I guess you now have to create a PR in the spec to add the `pulsar` binding and it will make it to the January release :)

**Pull request merged**
